### PR TITLE
MinGW-w64: update libcompat structure

### PR DESCRIPTION
**Pull request by yadij**
Separate the Squid for Windows portability fixes for MinGW and
MSVC builds. Beginning a new MinGW-w64 specific port with
a clean starting base.

With these changes the libcompat portability library builds.
The rest of Squid remains mostly non-building.

### CODE REVIEW

**Comment by yadij** [comment]
The rationale for this approach is that there is a lot of code in the _SQUID_WINDOWS_ which was designed for Visual Studio, hacked to work with MinGW32. That code is very old now, we can/will convert it as needed, but a clean start will allow us to identify which parts are actually unnecessary.

File: compat/os/mingw.h
Comment (rousskov) [inline compat/os/mingw.h:9]: The code changes in this PR appear to treat MinGW as a stand-alone OS rather than just a Windows "flavor". If that is the design we are going forward with, then we should name this guard consistent with guards in all other similar OS portability headers: `SQUID_OS_MINGW_H`.

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for working on improving Windows support! Are you running Squid on Windows or know somebody else who has tested these changes? Or are you relying on cross-build results, without actually running built Squid?

**Comment by kinkie** [comment]
How does this interact with https://github.com/kinkie/squid/tree/fix-mingw-cross-2 ? Should I abandon that work?

File: compat/os/mingw.h
Comment (yadij) [reply compat/os/mingw.h:19]: This is compliance with the rest of the portability files. Changing policy here is out of scope for this PR.

FWIW, the statement is clear. **everything** below that line is OS-specific and cannot be safely changed without expert knowledge of the POSIX API, the mapped system API, and active testing on the relevant OS build environment.

File: compat/statvfs.cc
Comment (yadij) [reply compat/statvfs.cc:24]: This "breaks" VisualStudio builds. But that environment has not built since Squid-2.7, with no support in sight it is effectively dead. So, I am switching these code blocks that are useful to MinGW.  When the MinGW port is working we can decide whether to drop the remainder VisualStudio-only code or not.

File: compat/os/mingw.h
Comment (yadij) [reply compat/os/mingw.h:12]: If one pays attention to the first word of the title (`Initial`) meaning start/beginning not the end/completion.

Comment (yadij) [reply compat/os/mingw.h:12]: Code in libcompat should (only) be added after build and testing using the environment+toolchain identifies a real need. The followup PRs in this project will be adding the portability hacks back as and when necessary based on fresh builds @kinkie and I are doing.

**Comment by yadij** [comment]
> > How does this interact with https://github.com/kinkie/squid/tree/fix-mingw-cross-2 ? Should I abandon that work?
> 
> FWIW, I agree that these are important questions. Please clarify this PR relationship with #1085.

I am hoping these two can work together, with this approach offering less work fixing broken hacks that are not actually necessary for MinGW. You should be able to rebase the `fix-mingw-cross-2` branch on this or master after this merges. This is in part why I scoped this PR as only libcompat.

File: compat/statvfs.cc
Comment (yadij) [reply compat/statvfs.cc:24]: In relation to PR #1085, we should copy the fixes from that PR code to this new `os/mingw.h` (when proven needed) instead of from the broken master branch `os/mswindows.h` file.

File: compat/os/mingw.h
Comment (yadij) [reply compat/os/mingw.h:12]: Is the new title better?

Comment (kinkie) [reply compat/os/mingw.h:12]: yes, thanks. It gives more clarity about the intended scope

File: compat/osdetect.h
Comment (yadij) [reply compat/osdetect.h:69]: All libcompat code that is currently inside `_SQUID_WINDOWS_` guards is **theoretically** dead code. ie it has not (yet) had a proven need to exist.

BUT, I do not dare drop it completely at this time since there are many parts of Squid still to be fixed which might prove a need and involve corrections to those wrappers.

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for addressing one of my concerns.

File: compat/os/mingw.h
Comment (yadij) [reply compat/os/mingw.h:12]: There is no more that can be said with any certainty. The named library **builds**, the rest is unknowable due to other code issues.

Comment (yadij) [reply compat/os/mingw.h:12]: Squid did not build with MinGW-w64 before this, and does not afterward either.

What we are getting here is a clean way to avoid the complexity and side effects @kinkie was faced with in PR #1085.

File: compat/os/mingw.h
Comment (yadij) [reply compat/os/mingw.h:19]: > > This is compliance with the rest of the portability files. Changing policy here is out of scope for this PR.
> 
> There is no such policy and no policy change is requested. 

This request is **creating** a new policy prohibiting this documentation.
Such policy change is out of scope for this PR.


> Not all portability files have this blob, so not adding it to this new file is not even going to create a new inconsistency.

All `compat/os/*.h` files should contain this warning. I see that Solaris and OpenSolaris broke the pattern. That will need fixing, but not here. This PR is maintaining consistency

> 
> > FWIW, the statement is clear. **everything** below that line is OS-specific and cannot be safely changed without expert knowledge of the POSIX API, the mapped system API, and active testing on the relevant OS build environment.
> 
> I am not against adding a statement similar to the above paragraph _if_ we want to block untested changes from non-experts. However, I doubt we want that
>. AFAICT, none of us would pass a meaningful bar in this particular context!

You are wrong. We do have the necessary expertise in the Squid Project.

> Let's document actual dependencies and other critical caveats (where it makes sense) rather than making grandiose statements about general dangers of modifying low-level OS-specific code.

No need for insults ("grandious"). This is not a general danger - this is very specific danger for these `compat/os/*.h` which need a very specific environment and toolchain and testing process to get any changes right. The same cannot be said of other Squid code.


> A lot of Squid code requires expert knowledge for serious modifications. Sprinkling code with such notices is likely to scare valuable contributions before we can review them.

In libcompat the definition of "valuable contribution" has somewhat stricter criteria than even the most complex other part of the Squid code. Making people reluctant to touch this particular code is the purpose.

File: compat/statvfs.cc
Comment (yadij) [reply compat/statvfs.cc:24]: > Why not remove all VisualStudio-only code as a starting point?

Several reasons:
 * I do not want to erase a lot of code only to re-add it.
 * Someone on the review team has a history of you rejecting such follow PRs on grounds of the re-addition being "new code". I do not want to spend years arguing over code style/logic/documentation which is currently status-quo just to get Squid to build again using the code it already has today.

Comment (yadij) [reply compat/statvfs.cc]: Reverted the removal in [ba0e1ea](https://github.com/squid-cache/squid/pull/1341/commits/ba0e1ea54eb845c4fb7326f3658b316fa3a94dae)

File: compat/osdetect.h
Comment (yadij) [reply compat/osdetect.h:69]: FTR, the [windows_ipaddrchangemonitor](http://www.squid-cache.org/Doc/config/windows_ipaddrchangemonitor/) directive has never worked in MinGW builds of Squid-3 by way of MinGW build never actually building the squid binary successfully. That feature and all other features are going to have build and run0-time testing in followup PRs. Code in src/ is out of scope for this PR.

Comment (yadij) [reply compat/osdetect.h:69]: The connection is that it is dead code either way. Code in src/ is out of scope for this PR.

**Comment by yadij** [comment]
@rousskov, please remove your veto and allow this to merge. The requests which I have not resolved are to be fixed in followup PRs.

File: compat/osdetect.h
Comment (rousskov) [inline compat/osdetect.h:68]: Since we are changing the semantics of this macro, it may be a good time to also fix the undefined behavior its name triggers. In [C++](https://en.cppreference.com/w/c/language/identifier), "All identifiers that begin with an underscore followed by a capital letter" ... "are reserved and may not be declared in a program (doing so invokes undefined behavior)". Those UB identifiers include macros, and `_SQUID_MINGW_` is one of them.

Removing the leading underscore will introduce and inconsistency with other similar OS detection `_SQUID_` macros, so I do not insist on that removal, but we may have enough excuses to introduce that inconsistency in this case. Your call.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: compat/osdetect.h
Comment (yadij) [reply compat/osdetect.h:68]: This macro is not being touched by this PR, nor is its meaning changed in any way. This request is out of scope.

File: compat/osdetect.h
Comment (yadij) [reply compat/osdetect.h:69]: > In summary, please make SQUID_WINDOWS and SQUID_MINGW guard usage consistent with their new mutually exclusive definition.

There are two places in the code which do not (yet) meet that criteria.
 * code in src/ which is out-of-scope for this PR. These are being resolved in followup PRs.
 * code in the `mswindows.{h,cc}` files. That is being left **for now** for the same reason all code inside other _SQUID_WINDOWS_ wrappers are being left as-is. Changing it requires the MSVC compiler to verify the change correctness.

File: compat/os/mingw.h
Comment (yadij) [reply compat/os/mingw.h:19]: Lets leave `os/mingw.h` matching the other operating system `os/*.h` files and save some brain cells.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: compat/os/mingw.h
Comment (yadij) [reply compat/os/mingw.h:19]: Indeed you are cycling. I have already addressed that seeming inconsistency.

File: compat/osdetect.h
Comment (yadij) [reply compat/osdetect.h:69]: You appear to have missed my point entirely. The requested change has far-reaching implications that I refuse to bring in-scope to this PR.
As much as you dislike the macro naming and usage it is the status-quo we have today. Please withdraw this request to change practice/policy.

Comment (rousskov) [reply compat/osdetect.h:69]: > The requested change has far-reaching implications

What far-reaching implications do you see?

> [the macro naming and usage] is the status-quo we have today

This PR changes the status quo:

* Before this PR, MinGW builds were a flavor or sub-category of Windows builds. MinGW builds used general Windows code (`_SQUID_WINDOWS_`) plus some MinGW-specific code.

* After this PR: MinGW builds are first-class or stand-alone builds like Linux or FreeBSD builds. They no longer use general Windows code (`_SQUID_WINDOWS_`).

The value of the `_SQUID_MINGW_` macro (i.e. 1) stays the same. The name of the macro guarding MinGW build is not of primary importance (to me) and can stay `_SQUID_MINGW_` if we want it to stay the same. However, this PR changes the semantics of what "MinGW build" means -- what code is enabled, what code is not enabled, what macros are defined, what macros are not defined, etc. Once this request is addressed, PR code will become consistent with that change in MinGW build semantics. Addressing this change request is important and straightforward. I have suggested specific ways to address it.

> Please withdraw this request to change practice/policy.

This change request does not request any practice or policy change. It only requests to remove an inconsistency introduced by PR changes.

Comment (rousskov) [reply compat/osdetect.h:69]: @yadij, please see compat/os/mswindows.h in master for an example where `_SQUID_MINGW_` is still used inside `_SQUID_WINDOWS_`.

```
upstream/master:compat/os/mswindows.h-17-#if _SQUID_WINDOWS_
upstream/master:compat/os/mswindows.h-18-
upstream/master:compat/os/mswindows.h-19-/****************************************************************************
upstream/master:compat/os/mswindows.h-20- *--------------------------------------------------------------------------*
upstream/master:compat/os/mswindows.h-21- * DO *NOT* MAKE ANY CHANGES below here unless you know what you're doing...*
upstream/master:compat/os/mswindows.h-22- *--------------------------------------------------------------------------*
upstream/master:compat/os/mswindows.h-23- ****************************************************************************/
upstream/master:compat/os/mswindows.h-24-
upstream/master:compat/os/mswindows.h-25-/* we target Windows XP and later - some API are missing otherwise */
upstream/master:compat/os/mswindows.h:26:#if _SQUID_MINGW_
```

Same for the current PR code.

I did not check for other occurrences. Please do.

Comment (kinkie) [reply compat/osdetect.h:69]: Would it be worth, in the future, to further unravel by declaring:

_SQUID_WINDOWS_ is code targeting the windows platform
_SQUID_MINGW_ is only code targetining mingw
_SQUID_MSVC_ or something similar targets specific oddities of the MSVC compiler

?

Comment (rousskov) [reply compat/osdetect.h:69]: > Amos: FTR, cleared up the misunderstanding during a dev Zoom meeting. The issue of _SQUID_MINGW_ use in code not touched by this PR is resolved.

Just to avoid a misunderstanding: This change request is still not addressed. During the meeting, you mentioned that the original problem went away because other PRs removed problematic "`_SQUID_MINGW_` inside `_SQUID_WINDOWS_`" code. I could not quickly check that assertion during the meeting, so I said something along the lines of "this change request is addressed provided that assertion is true (i.e. the problematic code is gone)".

After the meeting, I have checked that assertion: The problem is still there. After the meeting, I have [quoted](https://github.com/squid-cache/squid/pull/1341#discussion_r1243334118) some of the code that introduces the inconsistency that this change request is devoted to. I believe there are more such cases inside compat/os/mswindows.h (at least). If necessary, I can identify all such problematic places. I am happy to clarify anything that needs clarification, but the ball is on your side.

> Francesco: `_SQUID_WINDOWS_` is code targeting the windows platform

AFAICT, after this PR: `_SQUID_WINDOWS_` is code targeting the windows platform _with the exception_ of MinGW and CygWin builds. MinGW used to be included in `_SQUID_WINDOWS_`. This PR proposes to make the two mutually exclusive. There are no objections to that proposal, but the current PR code does not yet make the two mutually exclusive (from a human code reader point of view) -- we still have some `#if _SQUID_WINDOWS_ ... #if _SQUID_MINGW_` sections. This change request is about removing that inconsistency. There are several ways to remove it; some of the proposed solutions do not require a lot of work or a lot of changes.

**Review by rousskov — DISMISSED** [review]

File: compat/osdetect.h
Comment (yadij) [reply compat/osdetect.h:69]: Oh, those bits dropped in [f660d38](https://github.com/squid-cache/squid/pull/1341/commits/f660d3833d2f15404ee1a114b944598b93c44ebe). I was checking the src/ etc directories earlier.

**Pull request closed without merging**
